### PR TITLE
Add bucket index support to store gateway

### PR DESCRIPTION
**Pull request by pracucci**
**What this PR does**:
Following up the PR #3614, in this PR I'm adding bucket index support to the store-gateway. The usage of the bucket index in the store-gateway is optional and disabled by default.

I haven't changed any integration test because the existing tests in `integration/querier_test.go` already covers it.

**Which issue(s) this PR fixes**:
N/A

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/storegateway/bucket_index_metadata_fetcher.go [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+	markedForDeletionMeta = "marked-for-deletion"
+
+	// MarkedForNoCompactionMeta is label for blocks which are loaded but also marked for no compaction. This label is also counted in `loaded` label metric.
+	MarkedForNoCompactionMeta = "marked-for-no-compact"
+
+	// Modified label values.
+	replicaRemovedMeta = "replica-label-removed"
+)
+
+// fetcherMetrics is a copy of Thanos internal fetcherMetrics. These metrics have been copied from
```
Comment (pracucci) [inline pkg/storegateway/bucket_index_metadata_fetcher.go:164]: I've opened a PR in Thanos https://github.com/thanos-io/thanos/pull/3660 to export the `fetcherMetrics` to hopefully remove the duplicated code (even if may be tricky to integrate anyway due to the additional label values `corruptedBucketIndex` and `noBucketIndex`).

**Review by ranton256 — APPROVED** [review]
LGTM

**Comment by AllenzhLi** [comment]
Compactor block cleaner report an error when directory markers hava a block deletion mark which the block no longer exists.

```
component=cleaner org_id=fake msg="failed blocks cleanup and maintenance" err="marker not found"
```

**Comment by AllenzhLi** [comment]
> Compactor block cleaner report an error when directory markers hava a block deletion mark which the block no longer exists.
> 
> ```
> component=cleaner org_id=fake msg="failed blocks cleanup and maintenance" err="marker not found"
> ```

This will result in UpdateIndex always failure.

Maybe this should be a warning or just ignored.

**Comment by pracucci** [comment]
> > Compactor block cleaner report an error when directory markers hava a block deletion mark which the block no longer exists.
> > ```
> > component=cleaner org_id=fake msg="failed blocks cleanup and maintenance" err="marker not found"
> > ```
> 
> This will result in UpdateIndex always failure.
> 
> Maybe this should be a warning or just ignored.

Should be fixed by this change:
https://github.com/cortexproject/cortex/pull/3625/files#diff-821456401172c65d2c0cf8706a3d700bf213d73f6932cee9ef7d1d805693b097

If, once this PR is merge, it will happen again, then we can re-investigate it.

**Review by pstibrany — APPROVED** [review]
Excellent work, I don't see any flaws in it. Good job!

File: pkg/storage/tsdb/testutil/objstore.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+	"github.com/cortexproject/cortex/pkg/storage/bucket/filesystem"
+)
+
+func PrepareFilesystemBucket(t testing.TB) (objstore.Bucket, string) {
+	storageDir, err := ioutil.TempDir(os.TempDir(), "bucket")
+	require.NoError(t, err)
+
+	t.Cleanup(func() {
+		require.NoError(t, os.RemoveAll(storageDir))
```
[the remark this answers is not in the document]
Comment (pracucci) [reply pkg/storage/tsdb/testutil/objstore.go:19]: I would expect this cleanup to always work. If it doesn't, then there may be an underlying issue worth to investigate.

File: pkg/storegateway/bucket_index_metadata_fetcher.go [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+		// NOTE: filter can update synced metric accordingly to the reason of the exclude.
+		if customFilter, ok := filter.(MetadataFilterWithBucketIndex); ok {
+			err = customFilter.FilterWithBucketIndex(ctx, metas, idx, f.metrics.synced)
+		} else {
+			err = filter.Filter(ctx, metas, f.metrics.synced)
+		}
+
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "filter metas")
```
[the remark this answers is not in the document]
Comment (pracucci) [reply pkg/storegateway/bucket_index_metadata_fetcher.go:116]: I kept this logic as it was in Thanos metadata fetcher.

[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/storegateway/bucket_index_metadata_fetcher.go:116]: Discussed offline, Marco confirmed that it is ok to not call submit on each path, and we want to keep current API to match Thanos.

**Pull request merged**